packages/contracts-bedrock/src/L1/OPStackManager.sol

```diff
@@ -1,57 +1,327 @@
+        // assertValidContractAddress(address(_protocolVersions));
+        // assertValidContractAddress(_blueprints.addressManager);
+        // assertValidContractAddress(_blueprints.proxy);
+        // assertValidContractAddress(_blueprints.proxyAdmin);
+        // assertValidContractAddress(_blueprints.l1ChugSplashProxy);
+        // assertValidContractAddress(_blueprints.resolvedDelegateProxy);
+
+        superchainConfig = _superchainConfig;
+        protocolVersions = _protocolVersions;
+        blueprint = _blueprints;
+    }
+
+    /// @notice Callable by the OPSM owner to release a set of implementation contracts for a given
+    /// release version. This must be called with `_isLatest` set to true before any chains can be deployed.
+    /// @param _release The release version to set implementations for, of the format `op-contracts/vX.Y.Z`.
+    /// @param _isLatest Whether the release version is the latest released version. This is
+    /// significant because the latest version is used to deploy chains in the `deploy` function.
+    /// @param _setters The set of implementations to set for the release version.
+    function setRelease(string memory _release, bool _isLatest, ImplementationSetter[] calldata _setters) external {
+        // TODO Add auth to this method.
+
+        if (_isLatest) latestRelease = _release;
+
+        for (uint256 i = 0; i < _setters.length; i++) {
+            ImplementationSetter calldata setter = _setters[i];
+            Implementation storage impl = implementations[_release][setter.name];
+            if (impl.logic != address(0)) revert AlreadyReleased();
+
+            impl.initializer = setter.info.initializer;
+            impl.logic = setter.info.logic;
+        }
+    }
+
+    function deploy(DeployInput calldata _input) external returns (DeployOutput memory) {
+        assertValidInputs(_input);
+
+        // TODO Determine how we want to choose salt, e.g. are we concerned about chain ID squatting
+        // since this approach means a chain ID can only be used once.
+        uint256 l2ChainId = _input.l2ChainId;
+        bytes32 salt = bytes32(_input.l2ChainId);
+        DeployOutput memory output;
+
+        // -------- Deploy Proxy Contracts --------
+
+        // The ProxyAdmin is the owner of all proxies for the chain. We temporarily set the owner to
+        // this contract, and then transfer ownership to the specified owner at the end of deployment.
+        // The AddressManager is used to store the implementation for the L1CrossDomainMessenger
+        // due to it's usage of the legacy ResolvedDelegateProxy.
+        output.addressManager = AddressManager(Blueprint.deployFrom(blueprint.addressManager, salt));
+        output.opChainProxyAdmin =
+            ProxyAdmin(Blueprint.deployFrom(blueprint.proxyAdmin, salt, abi.encode(address(this))));
+        output.opChainProxyAdmin.setAddressManager(output.addressManager);
```

Comment:
Should this be moved above line 203? These technically aren't proxy contracts.

Reply:
Done in 0c40a9ea9